common/cross_sockets/XSocketServer.cpp

```diff
@@ -0,0 +1,134 @@
+#include "XSocketServer.h"
+
+#include "third-party/fmt/core.h"
+
+#include "common/cross_sockets/XSocket.h"
+
+#ifdef _WIN32
+#define NOMINMAX
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#include <WinSock2.h>
+#include <WS2tcpip.h>
+#endif
+
+XSocketServer::XSocketServer(std::function<bool()> shutdown_callback,
+                             int _tcp_port,
+                             int _buffer_size)
+    : want_exit_callback(std::move(shutdown_callback)) {
+  tcp_port = _tcp_port;
+  buffer_size = _buffer_size;
+  buffer = new char[_buffer_size];
```

Comment:
the codacy warning here is right - if you're allocating stuff in the constructor, you need to free it in the destructor and do something about copy constructor and assignment operator.

My suggestion is:
- replace buffer with `std::vector<char>`. If you use a vector, you don't have to worry about freeing the memory.  In the constructor, you can just do `buffer.resize(_buffer_size);`
- add this to the `.h`, inside `class XSocketServer`:
```cpp
XSocketServer(const XSocketServer&) = delete;
XSocketServer& operator=(const XSocketServer&) = delete;
```

This prevents somebody from accidentally copying the server object. This would be bad because both copies would hold the same socket, and then when the copies were eventually destroyed, they would both try to close the same socket.  If we add `=delete`, it will make it a compile time error.

Reply:
this was an error I had in my original `Listener` too.